asmcomp/linscan.ml

```diff
@@ -41,42 +53,73 @@ let rec insert_interval_sorted i = function
   | j :: _ as il when j.iend <= i.iend -> i :: il
   | j :: il -> j :: insert_interval_sorted i il
 
-let rec release_expired_fixed pos = function
-    i :: il when i.iend >= pos ->
-      Interval.remove_expired_ranges i pos;
-      i :: release_expired_fixed pos il
-  | _ -> []
-
-let rec release_expired_active ci pos = function
-    i :: il when i.iend >= pos ->
-      Interval.remove_expired_ranges i pos;
-      if Interval.is_live i pos then
-        i :: release_expired_active ci pos il
-      else begin
-        ci.ci_inactive <- insert_interval_sorted i ci.ci_inactive;
-        release_expired_active ci pos il
-      end
-  | _ -> []
-
-let rec release_expired_inactive ci pos = function
-    i :: il when i.iend >= pos ->
-      Interval.remove_expired_ranges i pos;
-      if not (Interval.is_live i pos) then
-        i :: release_expired_inactive ci pos il
-      else begin
-        ci.ci_active <- insert_interval_sorted i ci.ci_active;
-        release_expired_inactive ci pos il
-      end
-  | _ -> []
+let release_expired_spilled ci pos =
+  let (expired, divider_in_set, rest) =
+    (* (-1) is strictly below all valid positions, so (pos, (-1)) splits the set
+       into all spills with (iend < pos) and all spills above (iend >= 0),
+       without being itself in the set *)
+    SpilledSet.split (pos, (-1)) ci.ci_spilled in
+  assert (not divider_in_set);
+  ci.ci_free_slots <-
+    SpilledSet.fold (fun (_, ss) free -> IntSet.add ss free)
+      expired ci.ci_free_slots;
+  ci.ci_spilled <- rest
+
+let release_expired_fixed ci pos =
+  let rec loop ci pos = function
+      i :: il when i.iend >= pos ->
+        Interval.remove_expired_ranges i pos;
+        i :: loop ci pos il
```

Comment:
There are a few instances of non-tail recursive calls of the form `i :: loop ...`. If the recursion becomes deep enough, they could benefit from a `tail_mod_cons` annotation. Is it worth exploring? (I'm not familiar enough with the code to understand how likely it is we'll get deep recursion here).

Reply:
Personally I would rather consider rewriting these functions to use an `IntervalSet.t` instead of ordered list. This provides stack-space safety, but also better worst-case complexity. (For example I suspect that `release_expired_active` is currently quadratic, due repeated linear sorted insertion.) Both those possible suggestions could wait for a follow-up PR, though -- if @nojb is motivated enough.

Reply:
Wish granted! It needs a bit of care because `Interval.t` is mutable, but I believe the change is sound; review of the last commit would be welcome.